featurebyte/service/feature_table_cache_metadata.py

```diff
@@ -57,64 +63,113 @@ def __init__(
         )
         self.observation_table_service = observation_table_service
 
+    async def get_cached_definitions(
+        self, observation_table_id: PydanticObjectId
+    ) -> List[CachedDefinitionWithTable]:
+        """
+        Get cached feature definitions for observation table.
+
+        Parameters
+        ----------
+        observation_table_id: PydanticObjectId
+            Observation table id
+
+        Returns
+        -------
+        List[CachedFeatureDefinition]
+            Cached feature definitions
+        """
+        cached_definitions = []
+        async for cache_metadata in self.list_documents_iterator(
+            query_filter={"observation_table_id": observation_table_id}
+        ):
+            for feature in cache_metadata.feature_definitions:
+                cached_definitions.append(
+                    CachedDefinitionWithTable(
+                        feature_id=feature.feature_id,
+                        definition_hash=feature.definition_hash,
+                        feature_name=feature.feature_name,
+                        table_name=cache_metadata.table_name,
+                    )
+                )
+        return cached_definitions
+
     async def get_or_create_feature_table_cache(
         self,
         observation_table_id: PydanticObjectId,
+        num_columns_to_insert: int,
     ) -> FeatureTableCacheMetadataModel:
         """Get or create feature table cache document for observation table.
 
         Parameters
         ----------
         observation_table_id: PydanticObjectId
             Observation table id
+        num_columns_to_insert: int
+            Number of columns to insert
 
         Returns
         -------
         FeatureTableCacheMetadataModel
             Feature Table Cache model
         """
-        documents = []
-
         query_filter = {"observation_table_id": observation_table_id}
-        async for document in self.list_documents_iterator(query_filter=query_filter):
-            documents.append(document)
 
-        if documents:
-            document = documents[0]
-        else:
+        eligible_cache_metadata = None
+        num_cache_tables = 0
+        async for cache_metadata in self.list_documents_iterator(query_filter=query_filter):
+            if (
+                len(cache_metadata.feature_definitions) + num_columns_to_insert
```

Comment:
`num_columns_to_insert` should be much smaller than `FEATUREBYTE_FEATURE_TABLE_CACHE_MAX_COLUMNS` right?

Reply:
Currently `num_columns_to_insert` depends on the number of uncached features in a single request.

Usually it should be smaller than `FEATUREBYTE_FEATURE_TABLE_CACHE_MAX_COLUMNS`, but when it is larger the current implementation won't split the table but instead will create a new table that has more columns than `FEATUREBYTE_FEATURE_TABLE_CACHE_MAX_COLUMNS`. This shouldn't break anything unless `num_columns_to_insert` is significantly larger than the limit e.g. by 10x.

We can perform batching to ensure `num_columns_to_insert` is not too large as an improvement.

Reply:
I guess it should be fine since it will be a larger table than expected but does not need to be merged into, which is what we are trying to avoid here